feat(server_openvr): Add Quest Pro controller and headset emulation; Use VRLink hand tracking icons for the hand tracker controllers; Misc refactoring to reduce prop.rs linecount

shinyquagsire23 opened the pull request.

zmerp commented on alvr/server_core/src/connection.rs:
```diff
@@ -73,14 +73,18 @@ pub fn contruct_openvr_config(session: &SessionConfig) -> OpenvrConfig {
-            ControllersEmulationMode::ViveWand => 4,
-            ControllersEmulationMode::ViveTracker => 5,
-            ControllersEmulationMode::Custom { .. } => 6,
+            ControllersEmulationMode::QuestPro => 3,
+            ControllersEmulationMode::ValveIndex => 20,
+            ControllersEmulationMode::ViveWand => 40,
+            ControllersEmulationMode::ViveTracker => 41,
+            ControllersEmulationMode::Custom { .. } => 500,
```
Gaps are really useless lol. Soon I will implement schema-guided selection of the steamvr-restart settings. all the fields with leading underscore will be removed. I will add a single u64 at the root or the session which will be all the steamvr-restart settings hashed together.

shinyquagsire23 replied:
It's definitely more aesthetics/diff focused, either new enums get random numbers out of order or there's a bunch of extra diffs for new devices, so the gaps are marginally more friendly I guess

zmerp replied:
in any case, I don't have complaints since the code will go away soon

zmerp commented on alvr/server_openvr/src/props.rs:
```diff
@@ -577,6 +505,124 @@ pub extern "C" fn set_device_openvr_props(device_id: u64) {
                 ControllersEmulationMode::Custom { .. } => {}
             }
 
+            // Match for icons, to avoid duplicating across every Meta headset
+            // and also to override for the hand tracker controllers specifically
```
I don't think the approach is good. you can maybe make a function (even a nested function) and call it multiple times. Index controllers should have their own icons.

shinyquagsire23 replied:
tbh I'm not sure why Index controllers don't have icons specified. Like, they actually do show up fine, but they weren't specified before either.

zmerp replied:
Noone set them. This file is just full of random contributions

zmerp replied:
You probably have seen this issue #262

shinyquagsire23 replied:
oh right, I can throw some more accurate details in after I refactor then

zmerp reviewed (commented):
I wish we could merge my #2480 before, the props.rs is going to be a mess to merge... Sorry I didn't tell you before, I forgot about that PR.

shinyquagsire23 commented:
OK so,
- Refactored to the new stuff, and fixed a bug while I was at it (last PR accidentally set all info to the headset device ID lol)
- Added the VRLink stuff incl the model and input profile. Needs more work in a different PR, the shuttlecock is at the wrong position and none of the bindings work. But it's not any worse than it was tbh, adds a nice visual indication of hand tracking.
- Created a local function for all the icons stuff and moved everything to use it incl the headset emulation

zmerp commented on alvr/server_openvr/src/props.rs:
```diff
@@ -270,30 +268,24 @@ pub extern "C" fn set_device_openvr_props(device_id: u64) {
-                    "{htc}/icons/vive_headset_status_ready.png",
-                );
-                set_prop(
-                    NamedIconPathDeviceReadyAlertString,
-                    "{htc}/icons/vive_headset_status_ready_alert.png",
-                );
-                set_prop(
-                    NamedIconPathDeviceStandbyString,
-                    "{htc}/icons/vive_headset_status_standby.png",
+                    RegisteredDeviceTypeString,
+                    format!("oculus/{headset_serial}").as_str(),
                 );
+                set_icons("{oculus}/icons/quest_headset");
+            }
+            HeadsetEmulationMode::Vive => {
+                set_icons("{htc}/icons/vive_headset");
```
Move this back to the previous match

zmerp commented on alvr/server_openvr/src/props.rs:
```diff
@@ -654,53 +477,94 @@ pub extern "C" fn set_device_openvr_props(device_id: u64) {
-                        NamedIconPathDeviceNotReadyString,
-                        "{htc}/icons/tracker_status_error.png",
-                    );
+            // Match for icons, to avoid duplicating across every Meta headset
+            // and also to override for the hand tracking controllers specifically
+            match config.emulation_mode {
+                ControllersEmulationMode::RiftSTouch
+                | ControllersEmulationMode::Quest2Touch
+                | ControllersEmulationMode::Quest3Plus
+                | ControllersEmulationMode::QuestPro => {
+                    set_prop(TrackingSystemNameString, "oculus");
+
+                    set_prop(ControllerTypeString, "oculus_touch");
+                    set_prop(InputProfilePathString, "{oculus}/input/touch_profile.json");
+                    if left_hand {
+                        set_prop(
+                            RegisteredDeviceTypeString,
+                            format!("oculus/{headset_serial}_Controller_Left").as_str(),
+                        );
+                        set_icons("{oculus}/icons/rifts_left_controller");
+                    } else if right_hand {
+                        set_prop(
+                            RegisteredDeviceTypeString,
+                            format!("oculus/{headset_serial}_Controller_Right").as_str(),
+                        );
+                        set_icons("{oculus}/icons/rifts_right_controller");
+                    }
+                }
+                ControllersEmulationMode::ValveIndex => {
+                    if left_hand {
+                        set_icons("{indexcontroller}/icons/left_controller_status");
+                    } else if right_hand {
+                        set_icons("{indexcontroller}/icons/right_controller_status");
+                    }
+                }
+                // This is probably also a good switch statement as a fallback
```
Move this code back to the previous match and reduplicate some entries if necessary. It's not that I prefer duplicate entries but it makes things cleaner overall.

shinyquagsire23 replied:
I'll compromise and make it a closure instead

zmerp replied:
Yes you can make a reusable closure, but merge the match blocks.

shinyquagsire23 replied:
oh I already did that, check the new diff

shinyquagsire23 commented on alvr/server_openvr/src/props.rs:
```diff
@@ -197,71 +199,67 @@ pub extern "C" fn set_device_openvr_props(device_id: u64) {
         );
     };
 
+    let set_icons = |base_path: &str| {
+        set_prop(
+            NamedIconPathDeviceOffString,
+            format!("{base_path}_off.png").as_str(),
+        );
+        set_prop(
+            NamedIconPathDeviceSearchingString,
+            format!("{base_path}_searching.gif").as_str(),
+        );
+        set_prop(
+            NamedIconPathDeviceSearchingAlertString,
+            format!("{base_path}_searching_alert.gif").as_str(),
+        );
+        set_prop(
+            NamedIconPathDeviceReadyString,
+            format!("{base_path}_ready.png").as_str(),
+        );
+        set_prop(
+            NamedIconPathDeviceReadyAlertString,
+            format!("{base_path}_ready_alert.png").as_str(),
+        );
+        set_prop(
+            NamedIconPathDeviceAlertLowString,
+            format!("{base_path}_ready_low.png").as_str(),
+        );
+        set_prop(
+            NamedIconPathDeviceStandbyString,
+            format!("{base_path}_standby.png").as_str(),
+        );
+        set_prop(
+            NamedIconPathDeviceStandbyAlertString,
+            format!("{base_path}_standby_alert.gif").as_str(),
+        );
+    };
```
For whatever reason the compiler didn't like a function, something about it using the other internal function idk

shinyquagsire23 replied:
Actually yeah `set_prop` is a closure that reads `device_id` so no can do I'm afraid.

zmerp replied:
Aah right. I forgot. Github highlighting didn't give me the hint.

zmerp reviewed (approved):
Thanks for the set_openvr_prop fix. If you tested everything, looks good to me
